Bluetooth: Controller: Reduce assertion check code size

cvinayak opened the pull request:
Introduce development and fatal assertion classification in the Controller implementation.

Reduce Controller assertion check code size for ARM Cortex-M CPUs by raising undefined instruction exception or use faulting instruction address in the HCI vendor-specific h/w error event.

hci_ipc for nRF5340 Audio DK with nRF21540 FEM:
```
cmake -GNinja -DBOARD=nrf5340_audio_dk/nrf5340/cpunet -DCONF_FILE="nrf5340_cpunet_iso-bt_ll_sw_split.conf" -DEXTRA_CONF_FILE="overlay-nrf5340_cpunet_iso_nrf21540_ek-bt_ll_sw_split.conf" -DDTC_OVERLAY_FILE="boards/nrf5340_audio_dk_nrf5340_cpunet_nrf21540_ek.overlay" -DSNIPPET="bt-ll-sw-split" ../../samples/bluetooth/hci_ipc

```

Upstream memory use:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      248176 B       256 KB     94.67%
             RAM:       64016 B        64 KB     97.68%
           SRAM1:          0 GB        64 KB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
```

PR memory use when Controller assertion check use ARM Cortex-M faulting instruction address (and weirdly enabling HCI vendor-specific error event reduces ~8KB):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      213980 B       256 KB     81.63%
             RAM:       64340 B        64 KB     98.18%
           SRAM1:          0 GB        64 KB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
```

33.39 KB or ~ 13.78% reduction in code size. 

Test/example Controller assertion:
```
> HCI Event: Vendor (0xff) plen 38
        01 24 00 00 00 01 1e 00 00 00 01 00 00 00 00 00  .$..............
        00 00 0c bc 0a 00 14 37 00 00 a1 8b 01 00 ae 8b  .......7........
        01 00 9a 00 00 81                                ......          
```
where `ae 8b 01 00` is the program counter register value in little-endian octets.
or
```
***** USAGE FAULT *****
  Attempt to execute undefined instruction
r0/a1:  0x00000029  r1/a2:  0x00000001  r2/a3:  0x00000000
r3/a4:  0x00000745 r12/ip:  0x00000003 r14/lr:  0x00008b71
 xpsr:  0x8100009a
Faulting instruction address (r15/pc): 0x00008b80
```

Commandline to use to get details:
```
arm-none-eabi-addr2line -e <elf file path> <faulting instruction address>
```
Example:
```
$ arm-none-eabi-addr2line -e zephyr/zephyr.elf 0x00008b80
/home/jack/workspace/zephyrproject/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:233 (discriminator 1)
```

When using LTO, `arm-none-eabi-addr2line` is not able to fetch the source file and line number. With LTO, `.lst` file has to be used to look up the faulting instruction address. Hence, enable `CONFIG_OUTPUT_DISASSEMBLY=y` to get the `.lst` generated for the sample.

Minimal `peripheral_hr`:
```
cmake -GNinja -DBOARD=nrf54l15dk/nrf54l15/cpuapp -DCONF_FILE=prj_minimal.conf -DEXTRA_CONF_FILE=overlay-bt_ll_sw_split-minimal.conf ../../samples/bluetooth/peripheral_hr
```
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       96860 B      1428 KB      6.62%
             RAM:       17744 B       188 KB      9.22%
        IDT_LIST:          0 GB        32 KB      0.00%
```

This is an increase for the minimal, with this PR assertion checks are not removed in the minimal sample where as before it was removed when logging/console/serial was not included.

cvinayak committed:
Bluetooth: Controller: Introduce LL_ASSERT_DBG/ERR

Introduce development and fatal assertion classification
in the Controller implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Reduce assertion check code size

Reduce Controller assertion check code size for ARM Cortex-M
CPUs by using the undefined instruction exception.

`arm-none-eabi-addr2line` commandline can be used to get the
source file and line number.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (changes requested):
Please get rid of the fixup commit :)

cvinayak commented on subsys/bluetooth/controller/hal/debug.h:
```diff
@@ -27,6 +27,20 @@ void bt_ctlr_assert_handle(char *file, uint32_t line);
+#define HAL_DEBUG_LL_ASSERT(x) LL_ASSERT(x)
+#endif
+
+#define LL_ASSERT_ERR(cond) HAL_DEBUG_LL_ASSERT(cond)
+#define LL_ASSERT_DBG(cond) HAL_DEBUG_LL_ASSERT(cond)
```
`LL_ASSERT_DBG` is something that someone put in comment as `/* Should never happen */`  code path, example a value that was assigned is checked later, etc. And i vaguely seem to remember, some unit tests are using the assertion check to cover code path, hence want to keep it as-is.

The `LL_ASSERT_ERR` is fatal, and shall not be disabled.

With the use of 2-byte assertion code in this PR, having a Kconfig to enable/disable the debug `LL_ASSERT_DBG` is just 1% difference, and not attractive. But yes, in the future it will allow easy addition of a Kconfig to enable/disable it.

Thalley replied:
I feel like I'm still missing something: The two macros are identical, so why would you use one over the other?

Thalley replied:
This doesn't seem like a good solution to me. If you want to have something that is configurable, that should be a Kconfig option, and not a macro that you have to manually change to reduce code size

Thalley replied:
> Do not want yet a new Kconfig of already > 300 of them in the Controller :-)

Yeah I get that - Not yet sure if I can/will approve this as is, as configuring the stack by manually modifying a macro seems like a step in the wrong direction of how things should be configured in Zephyr. Would be nice to get a 3rd set of eyes on this

cvinayak committed:
samples: Bluetooth: hci_uart: Reuse hci_vs_err_assert

Reuse hci_vs_err_assert implementation to generate the HCI
vendor-specific hardware error event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: hci_ipc: Enable HCI vendor-specific h/w error event

Enable HCI vendor-specific h/w error event generation in
samples and tests.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Enable OUTPUT_DISASSEMBLY_WITH_SOURCE with LTO

Enable OUTPUT_DISASSEMBLY_WITH_SOURCE when using LTO to help
find the assertion check in the source code corresponding to
the faulting instruction address.

arm-none-eabi-addr2line tool is not able to resolve the
source file and line number when using LTO.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: hci_uart: Make bt_ll_sw_split conf consistent

Make overlay-all-bt_ll_sw_split configs consistent with the
one used in hci_ipc for nRF5340 SoC.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -281,47 +283,48 @@ static int h4_send(struct net_buf *buf)
-		file += pos;
-		len -= pos;
-	}
+	buf = hci_vs_err_assert(file, line);
```
Won't this treat file and line as a file name and line number instead of a fixed string and a PC value?

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -1197,6 +1209,17 @@ config BT_CTLR_ASSERT_HANDLER
 	  and will be invoked whenever the controller code encounters
 	  an unrecoverable error.
 
+config BT_CTLR_ASSERT_OPTIMIZE_FOR_SIZE
+	bool "Assertions optimized for code size"
+	default y if CPU_CORTEX_M
```
Is this supported by all architectures, or should there be one or more `depends on` here?

Thalley commented on include/zephyr/bluetooth/hci_vs.h:
```diff
@@ -234,8 +232,10 @@ struct bt_hci_vs_fata_error_cpu_data_cortex_m {
 	uint32_t a4;
 	uint32_t ip;
 	uint32_t lr;
+	uint32_t pc;
```
Does this count as a public API change? 

If anyone was using this VS command before, something may change. Given that this is just an addition, it should not be harmful I think

Thalley commented:
> ## [![Quality Gate Failed](https://camo.githubusercontent.com/ce402bcc4cf818eda660128f0b27a891eb8c9f4e0d783303ea667cb0d4b07f95/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d6661696c65642d323070782e706e67)](https://sonarcloud.io/dashboard?id=zephyrproject-rtos_zephyr&pullRequest=90599) **Quality Gate failed**
> 
> Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [E Reliability Rating on New Code](https://sonarcloud.io/dashboard?id=zephyrproject-rtos_zephyr&pullRequest=90599) (required ≥ C)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=zephyrproject-rtos_zephyr&pullRequest=90599)
> 
> ![](https://camo.githubusercontent.com/0d8093b061da31b2dde5907c87c133451a9fed9ab208145ec3750960f13f438a/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6c696768745f62756c622d313670782e706e67) Catch issues before they fail your Quality Gate with our IDE extension ![](https://camo.githubusercontent.com/cf28d9441c2be84b83964b59542f8523d293ea1d96a9d3cea77e7962537f0175/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f736f6e61726c696e742d313670782e706e67) [SonarQube for IDE](https://www.sonarsource.com/products/sonarlint/features/connected-mode/?referrer=pull-request)

@cvinayak You should be able to log in via Github and see if these errors are actually blockers from this PR, or leftovers from previous versions. If the latter, you should be able to accept them as a way of tell Sonar that they will be fixed in the future and not part of this PR, if you don't want to make this PR any more complex

Thalley reviewed (dismissed):
I don't see any immediate issues with this
